### ISSUE TITLE
API Proposal: Thread.JoinAsync

### ISSUE THREAD

**Issue by fbrosseau** [dotnet/runtime#40574]
## Background and Motivation

To bridge the gap between code that does explicit thread management, and async code, it would be great to have Thread.JoinAsync, analogous to Thread.Join. 

Based on the recently-added Process.WaitForExitAsync, I am proposing a single form taking a CancellationToken.

**The following description applies to Windows, I have zero idea of the implementability in other environments.**

While it is possible to implement this currently through PInvoke, it requires quite a lot of setup. 
- Obtain a handle to the thread somehow (this is not in the public API surface of Thread, so one way is to DuplicateHandle the thread's own pseudohandle)
- Wrap that into a WaitHandle implementation
- Wrap ThreadPool.RegisterWaitForSingleObject into a task

It is obviously also possible to implement this with just managed constructs, by wrapping the thread's main function into a try-finally that would signal some event/TaskCompletionSource, but this needs to be done at every thread callsite.

I think it would make sense to provide a clean, easy API for it instead.

## Proposed API

```
namespace System.Threading
{
     public class Thread
    {
+       public Task JoinAsync(CancellationToken ct);
    }
}
```

## Usage Examples

```
var myThread = new Thread(...);
myThread.Start();
await myThread.JoinAsync(ct);
```

**Comment by Wraith2** [comment]
Isn't the suggested way of doing this to create a `TaskCompletionSource` on the sync thread that you want to wait for and let the async portion of the code wait on that?

**Comment by fbrosseau** [comment]
Correct, it is possible to have similar functionality by wrapping a thread's main body method in a try-finally and setting a TaskCompletionSource accordingly. This proposal is about the convenience of having that done automatically and robustly.
